mutual-information-feature-selection/titanic.py

```diff
@@ -0,0 +1,155 @@
+import itertools
+import os
+
+import matplotlib.pyplot as plt
+import matplotlib.colors as colors
+import numpy as np
+import pandas as pd
+
+# D-Wave Ocean tools
+import dimod
+import neal
+from dwave.embedding.chimera import find_clique_embedding
+from dwave.system import DWaveSampler, FixedEmbeddingComposite
+
+# Read the feature-engineered data into a pandas dataframe
+# Data obtained from http://biostat.mc.vanderbilt.edu/DataSets
+datapath = os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                        'data',
+                        'formatted_titanic.csv')
+dataset = pd.read_csv(datapath)
+
+
+# Define MI calculations
+def prob(dataset):
+    """Joint probability distribution P(X) for the given data."""
+
+    # bin by the number of different values per feature
+    num_rows, num_columns = dataset.shape
+    bins = [len(np.unique(dataset[:, ci])) for ci in range(num_columns)]
+
+    prob, _ = np.histogramdd(dataset, bins)
+    return prob / np.sum(prob)
+
+
+def shannon_entropy(p):
+    """Shannon entropy H(X) is the sum of P(X)log(P(X)) for probabilty
+    distribution P(X).
+    """
+    p = p.flatten()
+    return -sum(pi*np.log2(pi) for pi in p if pi)
+
+
+def conditional_shannon_entropy(p, *conditional_indices):
+    """Conditional Shannon entropy H(X;Y) = H(X,Y) - H(Y)."""
+
+    axis = tuple(i for i in np.arange(len(p.shape))
+                 if i not in conditional_indices)
+
+    return shannon_entropy(p) - shannon_entropy(np.sum(p, axis=axis))
+
+
+def mutual_information(prob, j):
+    """Mutual information between variables X and variable Y.
+
+    Calculated as I(X; Y) = H(X) - H(X|Y)."""
+
+    return (shannon_entropy(np.sum(prob, axis=j))
+            - conditional_shannon_entropy(prob, j))
+
+
+def conditional_mutual_information(p, j, *conditional_indices):
+    """Mutual information between variables X and variable Y conditional on variable Z.
+
+    Calculated as I(X;Y|Z) = H(X|Z) - H(X|Y,Z)"""
+
+    return (conditional_shannon_entropy(np.sum(p, axis=j), *conditional_indices)
+            - conditional_shannon_entropy(p, j, *conditional_indices))
+
+
+# Rank the MI between survival and every other variable
+scores = {}
+features = list(set(dataset.columns).difference(('survived',)))
+for feature in features:
+    scores[feature] = mutual_information(prob(dataset[['survived', feature]].values), 0)
+
+labels, values = zip(*sorted(scores.items(), key=lambda pair: pair[1], reverse=True))
+
+# Plot the MI between survival and every other variable
+plt.figure()
+ax1 = plt.subplot(1, 2, 1)
+ax1.set_title("Mutual Information")
+ax1.set_ylabel('MI Between Survival and Feature')
+plt.xticks(np.arange(len(labels)), labels, rotation=90)
+plt.bar(np.arange(len(labels)), values)
+
+# The Titanic dataset provides a familiar, intuitive example available in the public
+# domain. In itself, however, it is not a good fit for solving by sampling. Run naively on
```

Comment:
I would add a one-sentence explanation on why it's not a good fit instead of leaving up to the reader's guess.